src/firefly/js/visualize/HiPSUtil.js

```diff
@@ -0,0 +1,344 @@
+    const {width,height}=pv.viewDim;
+    if (!cc || !width || !height) return;
+
+    const pt1= cc.getWorldCoords( makeDevicePt(1,height/2));
+    const pt2= cc.getWorldCoords( makeDevicePt(width-1,height/2));
+    return (pt1 && pt2) ? computeDistance(pt1,pt2) : 180;
+}
+
+
+const HealpixCache = {
+
+    lastNside : 8,
+    staticCache : {corners: {nside8: []}},
+    dynamicCache : {},
+    hpxIdxCache : null,
+
+
+    init() {
+        // pre-compute corners position for nside=8
+        const hpxIdx = new HealpixIndex(8);
+        hpxIdx.init();
+        const npix = HealpixIndex.nside2Npix(8);
+        let corners;
+        for (let ipix=0; ipix<npix; ipix++) {
+            corners =  hpxIdx.corners_nest(ipix, 1);
+            HealpixCache.staticCache.corners.nside8.push(corners);
+        }
+        HealpixCache.hpxIdxCache = hpxIdx;
+    },
+
+    corners_nest(ipix, nside) {
+        if (nside===8) {
+            return HealpixCache.staticCache.corners.nside8[ipix];
+        }
+
+        if (nside!==HealpixCache.lastNside) {
+            HealpixCache.hpxIdxCache = new HealpixIndex(nside);
+            HealpixCache.hpxIdxCache.init();
+            HealpixCache.lastNside = nside;
+        }
+
+        return HealpixCache.hpxIdxCache.corners_nest(ipix, 1);
+
+    }
+};
+
+HealpixCache.init();
+
+
+
+
+
+
+
+/**
+ *
+ * @param norder
+ * @param {WorldPt} centerWp - center of visible area, coorindate system of this point should be same as the projection
+ * @param {number} fov - Math.max(width, height) of the field of view in degrees (i think)
+ * @return {Array}
+ */
+export function getVisibleHiPSCells (norder, centerWp, fov) {
+    const cells = []; // array to be returned
+    const spVec = new SpatialVector();
+    const nside = Math.pow(2, norder); // aTODO : to be modified
+    const npix = HealpixIndex.nside2Npix(nside);
+    let ipixCenter = null;
+
+    let pixList;
+    if (fov>80) {
+        pixList = [];
+        for (let ipix=0; ipix<npix; ipix++) pixList.push(ipix);
+    }
+    else {
+        const hpxIdx = new HealpixIndex(nside);
+        hpxIdx.init();
+        const spatialVector = new SpatialVector();
+        spatialVector.set(centerWp.x, centerWp.y);
+        let radius = fov/2;
+        // we need to extend the radius
+        if (fov>60) {
+            radius *= 1.6;
+        }
+        else if (fov>12) {
+            radius *=1.45;
+        }
+        else {
+            radius *= 1.1;
+        }
+
+        pixList = hpxIdx.queryDisc(spatialVector, radius*Math.PI/180.0, true, true);
+        // add central pixel at index 0
+        const polar = Utils.radecToPolar(centerWp.x, centerWp.y);
+        ipixCenter = hpxIdx.ang2pix_nest(polar.theta, polar.phi);
+        pixList.unshift(ipixCenter);
+    }
+
+
+    let ipix;
+    for (let ipixIdx=0, len=pixList.length; ipixIdx<len; ipixIdx++) {
+        ipix = pixList[ipixIdx];
+        if (ipix===ipixCenter && ipixIdx>0) {
+            continue;
+        }
```

Comment:
It might be more easier to remove the duplicate from the list right away, then checking every element.